plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * WooCommerce Product Forms Controller
+ */
+
+namespace Automattic\WooCommerce\Admin\Features\ProductBlockEditor;
+
+/**
+ * Handle retrieval of product forms.
+ */
+class ProductFormsController {
+
+	/**
+	 * Set up the product forms controller.
+	 */
+	public function init() { // phpcs:ignore WooCommerce.Functions.InternalInjectionMethod.MissingFinal, WooCommerce.Functions.InternalInjectionMethod.MissingInternalTag -- Not an injection.
+		// Migrate form templates after WooCommerce plugin update.
+		add_action( 'upgrader_process_complete', array( $this, 'migrate_form_templates' ), 10, 2 );
+	}
+
+	/**
+	 * Migrate form templates.
+	 *
+	 * @param \WP_Upgrader $upgrader The WP_Upgrader instance.
+	 * @param array        $hook_extra Extra arguments passed to hooked filters.
+	 * @return void
+	 */
+	public function migrate_form_templates( \WP_Upgrader $upgrader, array $hook_extra ) {
+		// Check if the action is an `update` or `install` action for a plugin.
+		if (
+			'install' !== $hook_extra['action'] &&
+			'update' !== $hook_extra['action'] ||
+			'plugin' !== $hook_extra['type']
+		) {
+			return;
+		}
+
+		$updated_plugins = $hook_extra['plugins'];
+
+		// Check if WooCommerce plugin was updated.
+		if ( ! in_array( 'woocommerce/woocommerce.php', $updated_plugins, true ) ) {
+			return;
+		}
+
+		/**
+		 * Allow extend the list of templates that should be auto-generated.
+		 *
+		 * @since 9.1.0
+		 * @param array $templates List of templates to auto-generate.
+		 */
+		$templates = apply_filters(
```

Comment:
Is that something we want to make public right now? I've had a similar dilemma in [this PR](https://github.com/woocommerce/woocommerce/pull/47762/files#diff-c7e5447e92a9e360c5c5374feb6520a2602996e1f7a470a3574ac0bf0071af3bR104), where I marked a hook as experimental. Not sure if we have a correct process for new hooks.

Reply:
It's okay, considering the whole feature is hidden behind a feature flag. It's required for extensibility purposes.